File: life-assisitant-ui-mobile/src/api/health.ts
/**
 * 健康 API
 * 路径前缀：/health
 *
 * SYNC-FROM-BACKEND: life-assisitant-api/internal/controller/health.go
 *                    life-assisitant-api/internal/model/dto/health.go
 * 契约文档：md/spec-20260919-v1/05-API规范.md（health 段）
 *
 * 后端端点（router.go）与权限点：
 *   GET    /health/overview      health:view    概览卡 + 引导态所需的一切
 *   GET    /health/calendar      health:view    月历标记（只返回有标记的日期）
 *   GET    /health/days          health:view    区间日记
 *   GET    /health/days/:date    health:view    单日详情
 *   PUT    /health/days/:date    health:write   写日记（整体覆盖，见下方语义）
 *   DELETE /health/days/:date    health:write   删日记（返回 204，无响应体）
 *   GET    /health/settings      health:view    设置（懒创建，永远有值）
 *   PATCH  /health/settings      health:write   部分更新设置
 *   POST   /health/setup         health:write   引导向导提交（一次性）
 *
 *   GET    /health/events        health:view    时间轴事件（单日或区间）
 *   POST   /health/events        health:write   追加一次记录
 *   PATCH  /health/events/:id    health:write   改一次记录（date / metric_key 不可改）
 *   DELETE /health/events/:id    health:write   删一次记录（返回 204）
 *
 * ⚠️ 分工（2026-09-19 起，不可逆）：
 *   身体指标（water / bbt / weight / sleep / bowel）→ **只走 events**（一天多次、不延续）
 *   经期块 period + 心情块 moods + 备注            → 仍走 PUT /health/days/:date（一天一张）
 *
 * 注意：
 *   - `request.ts` 的响应拦截器已就地脱壳，业务方法拿到的就是 data 本身；
 *   - 204 已在拦截器里转成 null，所以删除类方法的返回类型是 void；
 *   - 整体覆盖语义：PUT 没给的字段置「未记录」，period / moods 子块整体不传 = 不触碰既有行；
 *     合并逻辑封装在 src/stores/health.ts 的 saveDay，本文件只负责收发。
 */
import { http } from './request'
import type {
  CreateHealthEventReq,
  CreateHealthEventResp,
  HealthCalendarResp,
  HealthDayDetailResp,
  HealthOverviewResp,
  HealthSettings,
  HealthSetupReq,
  HealthSetupResp,
  ListHealthDaysResp,
  ListHealthEventsResp,
  PatchHealthEventReq,
  PatchHealthEventResp,
  PatchHealthSettingsReq,
  UpsertHealthDayReq,
  UpsertHealthDayResp,
} from './types'

export const healthApi = {
  /** 概览卡 + 引导态（date 可选，用于补记/跨时区场景） */
  overview(date?: string): Promise<HealthOverviewResp> {
    return http.get<HealthOverviewResp>('/health/overview', { params: date ? { date } : {} })
  },

  /** 月历标记（month 必填，YYYY-MM；只返回有标记的日期） */
  calendar(month: string): Promise<HealthCalendarResp> {
    return http.get<HealthCalendarResp>('/health/calendar', { params: { month } })
  },

  /** 区间日记（闭区间，跨度 ≤366 天） */
  listDays(start: string, end: string): Promise<ListHealthDaysResp> {
    return http.get<ListHealthDaysResp>('/health/days', { params: { start, end } })
  },

  /** 单日详情；后端无记录时 day 为 null（不是 404） */
  getDay(date: string): Promise<HealthDayDetailResp> {
    return http.get<HealthDayDetailResp>(`/health/days/${date}`)
  },

  /**
   * 写某天日记（整体覆盖）。
   * 合并（取回→改→整体提交）在 store 层完成，本方法只负责发完整请求。
   * period / moods 子块缺省不传 = 不触碰既有行。
   */
  upsertDay(date: string, data: Omit<UpsertHealthDayReq, 'date'>): Promise<UpsertHealthDayResp> {
    return http.put<UpsertHealthDayResp>(`/health/days/${date}`, { ...data, date })
  },

  /** 删除某天记录（204） */
  deleteDay(date: string): Promise<void> {
    return http.delete<void>(`/health/days/${date}`)
  },

  /** 取设置（懒创建，不会 404） */
  getSettings(): Promise<HealthSettings> {
    return http.get<HealthSettings>('/health/settings')
  },

  /** 部分更新设置 */
  patchSettings(data: PatchHealthSettingsReq): Promise<HealthSettings> {
    return http.patch<HealthSettings>('/health/settings', data)
  },

  /** 引导向导提交（一次性） */
  setup(data: HealthSetupReq): Promise<HealthSetupResp> {
    return http.post<HealthSetupResp>('/health/setup', data)
  },

  /* ==================== 时间轴事件 ==================== */

  /**
   * 时间轴事件列表。
   * date 与 (start, end) 二选一；两个都不给 → 后端默认查今天。
   * 单日查询时响应里带 summary（当天从 events 聚合的汇总）。
   */
  listEvents(params: {
    date?: string
    start?: string
    end?: string
    metric_key?: string
  }): Promise<ListHealthEventsResp> {
    return http.get<ListHealthEventsResp>('/health/events', { params })
  },

  /**
   * 追加一次记录。
   * ⚠️ time 不传 = 服务端当前时刻 —— 前端不要自己算（客户端时钟可能不准/跨时区）。
   */
  createEvent(data: CreateHealthEventReq): Promise<CreateHealthEventResp> {
    return http.post<CreateHealthEventResp>('/health/events', data)
  },

  /** 改一次记录（date / metric_key 不可改，要挪到别处请删除+新建） */
  patchEvent(id: string, data: Omit<PatchHealthEventReq, 'id'>): Promise<PatchHealthEventResp> {
    return http.patch<PatchHealthEventResp>(`/health/events/${id}`, data)
  },

  /** 删一次记录（204） */
  deleteEvent(id: string): Promise<void> {
    return http.delete<void>(`/health/events/${id}`)
  },
}

export default healthApi

File: life-assisitant-api/internal/controller/health.go
// Package controller 健康模块控制器
//
// 9 个 handler 对应 md/spec-20260919-v1/07 §1 的 8 个接口。
// 读挂 health:view、写挂 health:write、设置与引导挂 health:manage（见 router.go）。
//
// ⚠️ GET /health/days/:date 无记录时返回 200 + data:{day:null}，**不是 404** ——
// 浮层要打开一张空表，404 会让前端被迫写「忽略错误」的兜底分支。
package controller

import (
	"github.com/gogf/gf/v2/net/ghttp"

	"github.com/life-assistant/api/internal/consts/ecode"
	"github.com/life-assistant/api/internal/model/dto"
	"github.com/life-assistant/api/internal/response"
	"github.com/life-assistant/api/internal/service"
)

// HealthController 健康控制器
type HealthController struct{}

var Health = &HealthController{}

// Overview GET /api/v1/health/overview?date=
// 概览卡 + 月历摘要条所需的一切，一次拉齐
func (c *HealthController) Overview(r *ghttp.Request) {
	var req dto.HealthOverviewReq
	_ = r.Parse(&req)
	out, err := service.Health().Overview(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// Calendar GET /api/v1/health/calendar?month=YYYY-MM
// 月历标记（只返回有标记的日期）
func (c *HealthController) Calendar(r *ghttp.Request) {
	var req dto.HealthCalendarReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().Calendar(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// ListDays GET /api/v1/health/days?start=&end=
// 区间记录（补记 / 报告页 / 导出），跨度上限 366 天
func (c *HealthController) ListDays(r *ghttp.Request) {
	var req dto.ListHealthDaysReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().ListDays(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// GetDay GET /api/v1/health/days/:date
// 单日详情（打开浮层时用）；无记录 day = null，接口仍返回 200
func (c *HealthController) GetDay(r *ghttp.Request) {
	out, err := service.Health().GetDay(r.Context(), r.Get("date").String())
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// UpsertDay PUT /api/v1/health/days/:date
// 核心写接口（整体覆盖；事务内联动 health_days + period_days + mood_logs）
func (c *HealthController) UpsertDay(r *ghttp.Request) {
	var req dto.UpsertHealthDayReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	// 日期在路径上，body 里也可能带一份（前端图省事会两边都传）→ 路径优先
	req.Date = r.Get("date").String()
	out, err := service.Health().UpsertDay(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// DeleteDay DELETE /api/v1/health/days/:date（204，无响应体）
// 前端删除后需自行重拉 overview
func (c *HealthController) DeleteDay(r *ghttp.Request) {
	if err := service.Health().DeleteDay(r.Context(), r.Get("date").String()); err != nil {
		writeError(r, err)
		return
	}
	response.NoContent(r)
}

// GetSettings GET /api/v1/health/settings
// 懒创建，永远有值 → 永不 404
func (c *HealthController) GetSettings(r *ghttp.Request) {
	out, err := service.Health().GetSettings(r.Context())
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// PatchSettings PATCH /api/v1/health/settings
// 只传要改的字段
func (c *HealthController) PatchSettings(r *ghttp.Request) {
	var req dto.PatchHealthSettingsReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().PatchSettings(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// Setup POST /api/v1/health/setup
// 首次引导一次性提交。⚠️ 前端「跳过」时不要调这个接口。
func (c *HealthController) Setup(r *ghttp.Request) {
	var req dto.HealthSetupReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().Setup(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// ============================================================================
// 身体指标时间轴事件（2026-09-19 新增）
//
// ⚠️ 这是 water / bbt / weight / sleep / bowel 五个指标的**唯一写入路径**。
// PUT /health/days/:date 已经不再处理它们（一天多次、不延续，装不进「一天一张表单」）。
// ============================================================================

// ListEvents GET /api/v1/health/events?date= 或 ?start=&end=&metric_key=
// 时间轴明细；只给 date 时会附带当天汇总 summary
func (c *HealthController) ListEvents(r *ghttp.Request) {
	var req dto.ListHealthEventsReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().ListEvents(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// CreateEvent POST /api/v1/health/events
// 追加一次记录。time 不传时由服务端取当前时刻（不让前端传，客户端时钟可能不准）。
func (c *HealthController) CreateEvent(r *ghttp.Request) {
	var req dto.CreateHealthEventReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Health().CreateEvent(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// PatchEvent PATCH /api/v1/health/events/:id
// 改某一次记录；date 与 metric_key 不可改（要改就删除 + 新建）
func (c *HealthController) PatchEvent(r *ghttp.Request) {
	var req dto.PatchHealthEventReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	req.ID = r.Get("id").String()
	out, err := service.Health().PatchEvent(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// DeleteEvent DELETE /api/v1/health/events/:id
// 删某一次记录。⚠️ 不是 204 —— 要返回删完之后的当日汇总，省得前端二次请求。
func (c *HealthController) DeleteEvent(r *ghttp.Request) {
	out, err := service.Health().DeleteEvent(r.Context(), r.Get("id").String())
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}
